Dapr component schema

FullStackChef opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠️ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! -->

<!-- ⚠️ **Every PR** needs to have a linked issue and have previously been approved. PRs that don't follow this will be rejected. >

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Closes #459 & #507 **

<!-- Add an overview of the changes here -->

<!-- All details should be in the linked issue. Feel free to call out any outstanding differences here. -->

## PR Checklist

<!-- Please check if your PR fulfills the following requirements, and remove the ones that are not applicable to the current PR -->

- [X] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [X] Based off latest main branch of toolkit
- [X] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [ ] New integration
  - [ ] Docs are written
  - [ ] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Contains **NO** breaking changes
- [] Every new API (including internal ones) has full XML docs
- [X] Code follows all style conventions

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

## Other information

<!-- Please add any other information that might be helpful to reviewers. -->

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:
```diff
@@ -269,16 +291,24 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
 
+        foreach (var secret in secrets)
+        {
+            Environment.SetEnvironmentVariable(secret.Key, secret.Value);
```
In the current process? Why?

FullStackChef replied:
haha "because it works"

Setting these environment variables loads them against the command line is executed - if there's a better place to do so I'd be happy to change as this felt a little like a random place to do it. 

The dapr local.env secret store  pulls variables from the env.

davidfowl replied:
What about collisions? You're setting this on the app host process. Explain to me how this is supposed to work? The app calls the dapr side car for secrets, that sidecar is configured with a secret store component that gets secrets from the side cars env?

FullStackChef replied:
Collisions with other aspire vars - should totally prefix this. collisions with multiple sidecars... too bad name your parameters better 😬 

Yes it is setting it on the app host process - the app host is what spawns the dapr command line right? 

dapr local env secret store works by exposing environment variables as "secrets" - it's designed for local development only

1. We have component (pubsub.rabbitmq) that has a parameter resource for password 
2. when we start daprcomponentsecretannotation exist so we need to add secret store
3. we add secret store component and add it to the sidecar paths
4.  setting the environment vars here is similar to exporting a var on the command line prior to calling dapr run ...
5. pubsub component has access to secret store therefore connection to rabbitmq is achieved

(the daprclient in dotnet also has access to these variables - but the primary focus here is the component access
e.g

```bash
export rabbitmq-password=SuperSecretPassword123
dapr run --app-id myapp dotnet run
```

sidecar has secret store and pubsub components - pubsub component retrieves password for rabbitmq from secret store

FullStackChef replied:
> > The dapr local.env secret store pulls variables from the env.
> 
> Missed this. I see, then I would just env variables on the sidecar itself.

I tried this (or at least I thought I did) I will take another crack at it see if I can't get it done that way

davidfowl replied:
This should work.

FullStackChef replied:
Have updated - I think I had tried setting the vars on the resource rather than the sidecar accidentally which is why it wasn't working - setting on sidecar is working. ick code gone

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:
```diff
@@ -430,82 +468,87 @@ private async Task<IReadOnlyDictionary<string, string>> StartOnDemandDaprCompone
+        return await contentWriter(content).ConfigureAwait(false);
+    }
+
+    private string GetAppHostRelativePath(string componentName)
```
Aside, we're creating a place to store things like this in 9.1, the aspire store.

FullStackChef replied:
nice -Apphost relative? wanna point me to the code and I could update inline with that?

FullStackChef replied:
I took a look at the aspire store interface - but it looks like a place to write data to rather than a location for user defined data - the idea of the app relative path is to provide a consistent location for users to place dapr components relative to the project rather than in the userprofile directory

davidfowl commented on examples/dapr/CommunityToolkit.Aspire.Hosting.Dapr.AppHost/Program.cs:
```diff
@@ -5,18 +5,23 @@
 var stateStore = builder.AddDaprStateStore("statestore");
+
 var pubSub = builder.AddDaprPubSub("pubsub")
+                           .WithMetadata("password", rmq.Resource.PasswordParameter)
```
```suggestion
                 .WithMetadata("password", rmq.Resource.PasswordParameter)
```

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.Dapr/DaprComponentSecretAnnotation.cs:
```diff
@@ -0,0 +1,4 @@
+﻿using Aspire.Hosting.ApplicationModel;
+
+namespace CommunityToolkit.Aspire.Hosting.Dapr;
+internal record DaprComponentSecretAnnotation(string Key, string Value) : IResourceAnnotation;
```
nit:
```suggestion

internal record DaprComponentSecretAnnotation(string Key, string Value) : IResourceAnnotation;
```

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.Dapr/DaprComponentSchema.cs:
```diff
@@ -0,0 +1,138 @@
+    /// <summary>
+    /// The name of the secret
+    /// </summary>
+    public required string Name { get; init; } = default!;
```
I don't think you need default! if it's required.

FullStackChef replied:
Ah you're right - I needed it for a couple of the other properties in order to allow the parameter less constructor - I've removed these ones

davidfowl commented on src/CommunityToolkit.Aspire.Hosting.Dapr/DaprComponentSchema.cs:
```diff
@@ -0,0 +1,138 @@
+    {
+        return serializer.Serialize(this);
+    }
+    public static implicit operator DaprComponentSchema(string yamlContent)
+    {
+        return deserializer.Deserialize<DaprComponentSchema>(yamlContent);
+    }
```
This feels bit too cute 😄

FullStackChef replied:
I **think** I know what you mean 😅
```csharp
public static DaprComponentSchema FromYaml(string yamlContent) =>
    deserializer.Deserialize<DaprComponentSchema>(yamlContent);
```
Does this make you happier? and can you tell me why?  Is it just that you prefer code to be more explicit?

davidfowl replied:
You're specifically trying to deserialize YAML. Besides the parameter name of this method, the implicit cast makes that invisible and this just leads to makes things less type safe.

paule96 commented:
Maybe if it is not to big of a no scope of this PR we should already start deviding dapr sidecar and component lifecycles to make our lifes easier later.

So that we would have two LifecycleHooks and that everytime you have something like `WithReference` on a sidecare you wait for this resource to complete.

With this we will have an easier life if we need to consider the endpoints to generate the correct YAML files

FullStackChef commented:
@paule96 my preference would be not to increase the scope of the my PR and I think the concept of separating lifecycle hooks needs further thought /design

aaronpowell commented:
/stale-extend

aaronpowell commented:
/stale-extend

FullStackChef commented on src/CommunityToolkit.Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:
```diff
@@ -392,6 +428,12 @@ private async Task<IReadOnlyDictionary<string, string>> StartOnDemandDaprCompone
+        // If any of the components have secrets, we will add an on-demand secret store component.
+        if (onDemandComponents.Any(component => component.TryGetAnnotationsOfType<DaprComponentSecretAnnotation>(out var annotations) && annotations.Any()))
+        {
+            onDemandComponents.Add(new DaprComponentResource("secretstore", DaprConstants.BuildingBlocks.SecretStore));
```
I don't think it's that simple. As a developer when I'm working locally, I want to be able to reference aspire parameters in my dapr metadata. If I have set those parameters as secrets then a dapr integration should respect that and treat them as such without me having to specifically set up a component.

Especially because when my resource requires secrets aspire automatically adds these secrets to a key vault and then references the key vault

However - looking at my implementation - I have overlooked what should happen if the user has configured a secrets component which should take priority

FullStackChef replied:
I've looked at this again and it turns out I did consider a user defined secrets component - like the existing pubsub & state, the secretstore will infact only use the local env in the absence of a secretstore.yaml in the apphost relative component directory (apphost/.dapr/components) and the dapr default component directory (userprofile/.dapr/components)

FullStackChef commented on examples/dapr/CommunityToolkit.Aspire.Hosting.Dapr.AppHost/.dapr/components/pubsub.yaml:
this file is used by this line of the apphost
```csharp
var pubSub = builder.AddDaprPubSub("pubsub")
```
this is not new  - it's simply in a project relative location

FullStackChef commented on examples/dapr/CommunityToolkit.Aspire.Hosting.Dapr.AppHost/Program.cs:
```diff
@@ -1,22 +1,28 @@
+       .WithDaprSidecar();
 
 builder.AddProject<Projects.CommunityToolkit_Aspire_Hosting_Dapr_ServiceB>("serviceb")
+       .WithReference(pubSub)
        .WithDaprSidecar()
```
this is up to the developer the both work the same

FullStackChef commented on src/CommunityToolkit.Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:
```diff
@@ -418,9 +461,10 @@ private async Task<IReadOnlyDictionary<string, string>> StartOnDemandDaprCompone
                 string componentPath = await (component.Type switch
                 {
-                    DaprConstants.BuildingBlocks.PubSub => GetPubSubAsync(component, contentWriter, cancellationToken),
-                    DaprConstants.BuildingBlocks.StateStore => GetStateStoreAsync(component, contentWriter, cancellationToken),
-                    _ => throw new InvalidOperationException($"Unsupported Dapr component type '{component.Type}'.")
+                    DaprConstants.BuildingBlocks.PubSub => GetBuildingBlockComponentAsync(component, contentWriter, "pubsub.in-memory", cancellationToken), // NOTE: In memory component can only be used within a single Dapr application.
+                    DaprConstants.BuildingBlocks.StateStore => GetBuildingBlockComponentAsync(component, contentWriter, "state.in-memory", cancellationToken),
+                    DaprConstants.BuildingBlocks.SecretStore => GetBuildingBlockComponentAsync(component, contentWriter, "secretstores.local.env", cancellationToken),
+                    _ => GetComponentAsync(component, contentWriter, cancellationToken)
                 }).ConfigureAwait(false);
```
because the first three cases in the switch statement handle specific building blocks with local development defaults

FullStackChef commented:
Quickly added parent relationship to address https://github.com/CommunityToolkit/Aspire/issues/507
![image](https://github.com/user-attachments/assets/524bc120-779d-4b58-af07-b20f5bd581db)

aaronpowell commented:
@FullStackChef is this ready to be considered reviewable/mergable?

FullStackChef commented:
From my point of view, yes

aaronpowell reviewed (commented):
From a purely code perspective, it looks all good, but as I'm not across this problem space I'll delegate the "does it work" stuff to others.

But if there are no complaints over the next few days I'd take that as "ready to merge"

FullStackChef commented:
@aaronpowell  - yep keen to get this merged so I can start looking at #516 as well as start building out some more functionality

aaronpowell commented:
Can you remove the PublicAPI file and it'll be good to merge.

I think I'll add this with the Redis one, and call that 9.3 so we can ship the new integration - that sound good?

FullStackChef commented:
@aaronpowell sounds good - I've removed the public api files and fixed the conflict with the packages file. Once this is merged I'll try to get a PR for #516 later today so we can squeeze it into 9.3 if possible - it should be a pretty small fix (check for customization annotation and configure based appropriately based on ) but it represents a fairly blocking issue

aaronpowell commented:
ok, let's merge this in, that way you're unblocked but I won't release a 9.2.2, we'll have that in pre-release and go straight to 9.3.
